### PR TITLE
[ray client] ray wait() implementation

### LINKED ISSUE #11979: Support ray.wait

### PR DESCRIPTION
**Pull request by AmeerHajAli**
This PR implements the `ray.wait()` experimental function call for the ray client support.
closes #11979.
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by barakmich** [comment]
So this works, but it's worth thinking a bit about the API here. We have an opportunity to make it nice; I might suggest streaming a unary/stream RPC, and the the awaited results in any order . Sure, under the hood, for the moment, it's the same call -- the server part of this code is basically the same -- but the client can be a little more dynamic and do the associated ordering as things come back.

This does have implications for the `get()` service boundary -- how is `get()` different than a `wait()` with no timeout?

**Comment by ericl** [comment]
@barakmich can you file an issue to support streaming optimizations in the future? I don't think we should block this PR for now

**Comment by AmeerHajAli** [comment]
I agree with @ericl . I think this needs better optimization as `wait` without timeout is different than `get` in the sense that for `wait` to run you need all the references while for `get` you can easily stream it.

File: src/ray/protobuf/ray_client.proto [hunk 1]
```diff
@@ -56,12 +56,25 @@ message GetResponse {
   bool valid = 1;
   bytes data = 2;
 }
+message WaitRequest {
+  bytes object_refs = 1;
+  bytes num_returns = 2;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:61]: int64

File: src/ray/protobuf/ray_client.proto [hunk 2]
```diff
@@ -56,12 +56,25 @@ message GetResponse {
...
+  bytes timeout = 3;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:62]: int64

File: src/ray/protobuf/ray_client.proto [hunk 3]
```diff
@@ -56,12 +56,25 @@ message GetResponse {
...
+}
+
+message WaitResponse {
+  bool valid = 1;
+  bytes ready_object_ids = 2;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:67]: repeated bytes ready_object_ids

File: src/ray/protobuf/ray_client.proto [hunk 4]
```diff
@@ -56,12 +56,25 @@ message GetResponse {
...
+  bytes remaining_object_ids = 3;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:68]: repeated bytes remaining_object_ids

File: src/ray/protobuf/ray_client.proto [hunk 5]
```diff
@@ -56,12 +56,25 @@ message GetResponse {
   bool valid = 1;
   bytes data = 2;
 }
+message WaitRequest {
+  bytes object_refs = 1;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:60]: repeated bytes object_refs

File: src/ray/protobuf/ray_client.proto (hunk 2 above)
Comment (AmeerHajAli) [reply src/ray/protobuf/ray_client.proto]: This should be double right?

Comment (ericl) [reply src/ray/protobuf/ray_client.proto]: Gotcha

File: python/ray/experimental/client/worker.py [hunk 6]
```diff
@@ -59,6 +66,38 @@ def _put(self, val):
...
+             *,
+             num_returns: int = 1,
+             timeout: float = None
+             ) -> (List[ClientObjectRef], List[ClientObjectRef]):
+        assert isinstance(object_refs, list)
+        for ref in object_refs:
+            assert isinstance(ref, ClientObjectRef)
+        data = {
+            "object_refs": [
+                cloudpickle.dumps(object_ref) for object_ref in object_refs
```
Comment (ericl) [inline python/ray/experimental/client/worker.py:80]: Instead of encoding with cloudpickle, can we use the raw bytes here? `object_ref.binary()`

**Review by ericl — APPROVED** [review]
Just one request to use .binary()

File: python/ray/experimental/client/worker.py (hunk 6 above)
Comment (AmeerHajAli) [reply python/ray/experimental/client/worker.py:80]: Not really, note that the type of object ref is not ObjectRef but rather it is ClientObjectRef.

Comment (AmeerHajAli) [reply python/ray/experimental/client/worker.py:80]: We can do that on the server-side though. I changed the server-side to not dump the returned ids and hence we do not have to load them on the worker side.

**Comment by ericl** [comment]
Looks good pending tests.

**Comment by ericl** [comment]
Lint errors

**Pull request merged**